### PR TITLE
Adding loader for IMB's YFCC dataset

### PR DESCRIPTION
**Pull request by ayushbaid**
This PR adds a loader for the dataset provided by the [image-matching-benchmark](https://www.cs.ubc.ca/research/image-matching-challenge/data/). 

This dataset will be used to evaluate and benchmark frontends.

### CODE REVIEW

File: gtsfm/loader/yfcc_imb_loader.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                w=quaternions[0],
+                x=quaternions[1],
+                y=quaternions[2],
+                z=quaternions[3],
+            )
+
+            cTw = Pose3(rot, image_params[indices_map[idx]].tvec)
+
+            K_params = camera_params[indices_map[idx]].params
+            # TODO: cannot fit two focal lengths k_params[0] and k_params[1]
```
Comment (johnwlambert) [inline gtsfm/loader/yfcc_imb_loader.py:179]: thanks for the PR, Ayush. In the IMB ground truth, are there any cameras w/ nonsquare pixels? (fx != fy)

**Comment by johnwlambert** [comment]
I like the PR overall Ayush, but one concern is the disk space added to the repo. I believe the files are 57.747 MB, which will slow down the pull/push times quite a bit. Another strategy is to put 6 separate 10 MB zip files behind wget links, and add a `getdata.sh` script that `wgets` them.

**Comment by johnwlambert** [comment]
You could use these wget links instead:

https://github.com/borglab/gtsfm/files/5898320/sparse.zip
https://github.com/borglab/gtsfm/files/5898324/new-vis-pairs.zip
https://github.com/borglab/gtsfm/files/5898332/images1.zip
https://github.com/borglab/gtsfm/files/5898335/images2.zip
https://github.com/borglab/gtsfm/files/5898337/images3.zip

File: gtsfm/loader/yfcc_imb_loader.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import numpy as np
+from gtsam import Cal3Bundler, PinholeCameraCal3Bundler, Pose3, Rot3
+
+import gtsfm.utils.io as io_utils
+import thirdparty.colmap.read_write_model as colmap_read_write_model
+from gtsfm.common.image import Image
+from gtsfm.loader.loader_base import LoaderBase
+
+
+class YfccImbLoader(LoaderBase):
```
Comment (johnwlambert) [inline gtsfm/loader/yfcc_imb_loader.py:19]: is there a Github python file from the IMB Github we are basing some of this off of?

Comment (johnwlambert) [reply gtsfm/loader/yfcc_imb_loader.py:19]: Maybe we could add a ref to it, if there is

File: gtsfm/loader/yfcc_imb_loader.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    """Loader for IMB's YFCC dataset."""
+
+    def __init__(self, folder: str, coviz_thresh: float = 0.1):
+        """Initializes the loader.
+
+        Args:
+            folder: the base folder of the dataset.
+            coviz_thresh (optional): threshold for covisibility between two
```
Comment (akshay-krishnan) [inline gtsfm/loader/yfcc_imb_loader.py:27]: Read this as covid_thresh... oh no :(

File: gtsfm/loader/yfcc_imb_loader.py (hunk 1 above)
Comment (ayushbaid) [reply gtsfm/loader/yfcc_imb_loader.py:179]: Yup. the images are non-square

**Comment by ayushbaid** [comment]
> I like the PR overall Ayush, but one concern is the disk space added to the repo. I believe the files are 57.747 MB, which will slow down the pull/push times quite a bit. Another strategy is to put 6 separate 10 MB zip files behind wget links, and add a `getdata.sh` script that `wgets` them.

We are now at 5.1MB total for the dataset. Does this work?

File: tests/loader/test_yfcc_imb_loader.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        index_to_test = 5
+        file_path = FOLDER_PATH / "images" / "05866831_3427466899.jpg"
+
+        loader_image = self.loader.get_image(index_to_test)
+
+        expected_image = io_utils.load_image(file_path)
+
+        np.testing.assert_allclose(
+            expected_image.value_array, loader_image.value_array
+        )
```
Comment (akshay-krishnan) [inline tests/loader/test_yfcc_imb_loader.py:67]: nit and more of personal preference: Isn't that too many empty lines between code?

**Review by akshay-krishnan — APPROVED** [review]

File: gtsfm/loader/yfcc_imb_loader.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                                     Defaults to 0.1.
+        """
+
+        self._folder = folder
+
+        # load all the image pairs according to the covisibility threshold used
+        # in IMB's reporting
+        visibility_file = osp.join(
+            self._folder,
+            "new-vis-pairs",
```
Comment (johnwlambert) [inline gtsfm/loader/yfcc_imb_loader.py:38]: does the "new-vis-pairs" terminology come from IMB @ayushbaid ?

File: gtsfm/loader/yfcc_imb_loader.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/loader/yfcc_imb_loader.py:182]: What about the pixels themselves (not the images)?

If the pixels are square, then fx=fy, so the TODO could be removed. But pixels are usually square

File: gtsfm/loader/yfcc_imb_loader.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/loader/yfcc_imb_loader.py:19]: @ayushbaid bumping this comment

**Review by johnwlambert — APPROVED** [review]

**Comment by johnwlambert** [comment]
Overall looks good, 2 comments you could please address then we can merge

**Comment by johnwlambert** [comment]
@ayushbaid what do the keys* files (e.g. "data/imb_reichstag/new-vis-pairs/keys-th-0.1.npy ") represent?

File: gtsfm/loader/yfcc_imb_loader.py (hunk 5 above)
Comment (ayushbaid) [reply gtsfm/loader/yfcc_imb_loader.py:38]: yes. Its their variable naming convention. From what I understand, they did their own co-visibility check and are not relying on the original YFCC dataset's information

File: gtsfm/loader/yfcc_imb_loader.py (hunk 2 above)
Comment (ayushbaid) [reply gtsfm/loader/yfcc_imb_loader.py:19]: Thanks for this. I now notice there are actually two different places from which the camera pose information is read. I used this as a reference: https://github.com/vcg-uvic/image-matching-benchmark/blob/master/example/training_data/parse_data.ipynb. However, in the stereo computation, a different source is used: https://github.com/vcg-uvic/image-matching-benchmark/blob/d6953b32e07e3545ebd50ae52887cc7c91e433a1/compute_stereo.py#L77

**Comment by ayushbaid** [comment]
> @ayushbaid what do the keys* files (e.g. "data/imb_reichstag/new-vis-pairs/keys-th-0.1.npy ") represent?

These are visibility files generated by the IMB folks

**Pull request closed without merging**